drivers/regulator/regulator_axp192.c

```diff
@@ -0,0 +1,438 @@
+		LOG_INST_DBG(config->log, "[0x%x]=0x%x mask=0x%x", (uint32_t)config->desc->vsel_reg,
+			     idx, (uint32_t)config->desc->vsel_mask);
+		ret = i2c_reg_update_byte_dt(&cconfig->i2c, config->desc->vsel_reg,
+					     config->desc->vsel_mask, (uint8_t)idx);
+
+		if (ret != 0) {
+			LOG_INST_ERR(config->log, "Failed to set regulator voltage");
+			return ret;
+		}
+
+		/* make sure regulator is enabled */
+		ret = axp192_enable(dev);
+		if (ret != 0) {
+			return ret;
+		}
+	}
+
+	return ret;
+}
+
+int axp192_get_voltage(const struct device *dev, int32_t *volt_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_common_config *cconfig = config->mfd->config;
+	int ret;
+	uint8_t raw_reg;
+
+	/* check if regulator is disabled */
+	ret = i2c_reg_read_byte_dt(&cconfig->i2c, config->desc->enable_reg, &raw_reg);
+	if (ret != 0) {
+		return ret;
+	}
+
+	if (0u == (raw_reg & config->desc->enable_mask)) {
+		/* regulator is disabled */
+		*volt_uv = 0u;
+	} else {
+		/* regulator is enabled, read voltage */
+		ret = i2c_reg_read_byte_dt(&cconfig->i2c, config->desc->vsel_reg, &raw_reg);
+		if (ret != 0) {
+			return ret;
+		}
+
+		raw_reg = (raw_reg & config->desc->vsel_mask) >> config->desc->vsel_bitpos;
+
+		ret = linear_range_group_get_value(config->desc->ranges, config->desc->num_ranges,
+						   raw_reg, volt_uv);
+	}
+
+	return ret;
+}
+
+int axp192_get_current_limit(const struct device *dev, int32_t *curr_ua)
+{
+	const struct regulator_axp192_config *config = dev->config;
+
+	*curr_ua = config->desc->max_ua;
+
+	return 0;
+}
+
+static struct regulator_driver_api api = {
+	.enable = axp192_enable,
+	.disable = axp192_disable,
+	.count_voltages = axp192_count_voltages,
+	.list_voltage = axp192_list_voltage,
+	.set_voltage = axp192_set_voltage,
+	.get_voltage = axp192_get_voltage,
+	.get_current_limit = axp192_get_current_limit,
+};
+
+static int regulator_axp192_init(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_common_config *cconfig = config->mfd->config;
+	uint8_t enabled_val;
+	bool is_enabled;
+	int ret = 0;
+
+	regulator_common_data_init(dev);
+
+	LOG_INST_DBG(config->log, "voltage=%d", config->common.init_uv);
+
+	if (!device_is_ready(config->mfd)) {
+		LOG_INST_ERR(config->log, "Parent instance not ready!");
+		return -ENODEV;
+	}
+
+	/*
+	 * make sure DCDC1-3 are set to automatic mode (PWM/PFM). This can be be set
+	 * independet of actual enable state
+	 */
+	ret = i2c_reg_update_byte_dt(&cconfig->i2c, AXP192_REG_DCDC123_WORKMODE,
+				     (AXP192_DCDC1_WORKMODE_MASK | AXP192_DCDC2_WORKMODE_MASK |
+				      AXP192_DCDC3_WORKMODE_MASK),
+				     (AXP192_DCDC1_WORKMODE_AUTO | AXP192_DCDC2_WORKMODE_AUTO |
+				      AXP192_DCDC3_WORKMODE_AUTO));
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "Setting DCDC1-3 workmode failed");
+		return ret;
+	}
```

Comment:
why is mode set here? the API should support the mode set/get

Reply:
Most, if not all, regulators support multiple operating modes, so unless the vendor tells you that they can't be used, they should be supported by the API. I guess `AUTO` will be the default if not specified. Certain modes only make sense in low-power scenarios, but it's really the user who needs to decide (or like in Linux, optionally make an automatic choice based on the load specified by its clients)

Reply:
Corrected